### ISSUE TITLE
[MU4 Issue] Stack buffer overflow vulnerability while parse MIDI file

### ISSUE THREAD

**Issue by kunshim** [musescore/MuseScore#16346]
**Describe the bug**
It only affects the Windows version.
In src/importexport/midi/internal/midishared/midifile.cpp
```cpp
void MidiFile::skip(qint64 len)
{
    // Note: if MS is updated to use Qt 5.10, this can be implemented with QIODevice::skip(), which should be more efficient
    //       as bytes do not need to be moved around.
    if (len <= 0) {
        return;
    }
#if (!defined (_MSCVER) && !defined (_MSC_VER))
    char tmp[len];
    read(tmp, len);
#else
    const int tmp_size = 256;    // Size of fixed-length temporary buffer. MSVC does not support VLA.
    char tmp[tmp_size];
    while (len > tmp_size) {
        read(tmp, len); //vulnerability 
        len -= tmp_size;
    }
    // Now len is <= tmp_size, last read fits in the buffer.
    read(tmp, tmp_size);
#endif
}
```
It is copying the length of len, not tmp_size. If len is bigger than 256, it will overwrite stack. 
This leads to a buffer overflow and overwrite the stack cookies. If an attacker knows the stack cookie value, it can lead to code execution. Check PoC.midi in PoC.zip. 
This vulnerability works from version 3.0 to 4.0.1(latest)

![image](https://user-images.githubusercontent.com/20679212/218482864-4894105d-80ce-4fc7-8a81-9ac215116f4d.png)
[PoC.zip](https://github.com/musescore/MuseScore/files/10723083/PoC.zip)

If you have any problems, contact me via kunshim@naver.com

**Comment by Jojo-Schmitz** [comment]
How about this then:
```
#if (!defined (_MSCVER) && !defined (_MSC_VER))
      char tmp[len];
      read(tmp, len);
#else
      // MSVC does not support VLA. Replace with std::vector. If profiling determines that the
      //    heap allocation is slow, an optimization might be used.
      std::vector<char> tmp(len);
      read(&tmp, len);
#endif
```

**Comment by kunshim** [comment]
Pretty good!

**Comment by Jojo-Schmitz** [comment]
Another possible fix would be:
```#if (!defined (_MSCVER) && !defined (_MSC_VER))
    char tmp[len];
    read(tmp, len);
#else
    const int tmp_size = 256;    // Size of fixed-length temporary buffer. MSVC does not support VLA.
    char tmp[tmp_size];
    while (len > tmp_size) {
        read(tmp, tmp_size);
        len -= tmp_size;
    }
    // Now len is <= tmp_size, last read fits in the buffer.
    read(tmp, len);
#endif
}
```
I.e. in the loop real tmp_size bytes and after the loop read (the remaining) len bytes. I guess that was the original intention.

**Comment by Jojo-Schmitz** [comment]
Another possible fix would be:
```
#if (!defined (_MSCVER) && !defined (_MSC_VER))
    char tmp[len];
    read(tmp, len);
#else
    const int tmp_size = 256;    // Size of fixed-length temporary buffer. MSVC does not support VLA.
    char tmp[tmp_size];
    while (len > tmp_size) {
        read(tmp, tmp_size);
        len -= tmp_size;
    }
    // Now len is <= tmp_size, last read fits in the buffer.
    read(tmp, len);
#endif
}
```
I.e. in the loop read `tmp_size` bytes and after the loop read (the remaining) `len` bytes rather than the other way round. I guess that was the original intention, back then when we moved from MinGW to MSVC.
@anatoly-os?

**Comment by Jojo-Schmitz** [comment]
Good question...
So something like this:
```
void MidiFile::skip(qint64 len)
{
    if (len <= 0) {
        return;
    }
    fp->skip(len);
}
```

**Comment by Jojo-Schmitz** [comment]
Further simplified it, eliminating that function entirely

**Comment by kunshim** [comment]
CVE-2023-26923 assigned

**Comment by Jojo-Schmitz** [comment]
You may want to mention that is fixed in 4.0.2 and later and also in my 3.7 from https://github.com/Jojo-Schmitz/MuseScore/tree/3.x
